main.go

```diff
@@ -1 +1,150 @@
+	"github.com/metal3-io/baremetal-operator/pkg/version"
+	metal3iocontroller "github.com/openshift/image-customization-controller/controllers/metal3.io"
+	"github.com/openshift/image-customization-controller/pkg/imagehandler"
+	// +kubebuilder:scaffold:imports
+)
+
+var (
+	scheme   = k8sruntime.NewScheme()
+	setupLog = ctrl.Log.WithName("setup")
+)
+
+func init() {
+	_ = clientgoscheme.AddToScheme(scheme)
+
+	_ = metal3iov1alpha1.AddToScheme(scheme)
+	// +kubebuilder:scaffold:scheme
+}
+
+func printVersion() {
+	setupLog.Info(fmt.Sprintf("Go Version: %s", runtime.Version()))
+	setupLog.Info(fmt.Sprintf("Go OS/Arch: %s/%s", runtime.GOOS, runtime.GOARCH))
+	setupLog.Info(fmt.Sprintf("Git commit: %s", version.Commit))
+	setupLog.Info(fmt.Sprintf("Build time: %s", version.BuildTime))
+	setupLog.Info(fmt.Sprintf("Component: %s", version.String))
+}
+
+func setupChecks(mgr ctrl.Manager) {
+	if err := mgr.AddReadyzCheck("ping", healthz.Ping); err != nil {
+		setupLog.Error(err, "unable to create ready check")
+		os.Exit(1)
+	}
+
+	if err := mgr.AddHealthzCheck("ping", healthz.Ping); err != nil {
+		setupLog.Error(err, "unable to create health check")
+		os.Exit(1)
+	}
+}
+
+func main() {
+	var watchNamespace string
+	var devLogging bool
+	var imagesBindAddr string
+	var imagesPublishAddr string
+
+	// From CAPI point of view, BMO should be able to watch all namespaces
+	// in case of a deployment that is not multi-tenant. If the deployment
+	// is for multi-tenancy, then the BMO should watch only the provided
+	// namespace.
+	flag.StringVar(&watchNamespace, "namespace", os.Getenv("WATCH_NAMESPACE"),
+		"Namespace that the controller watches to reconcile host resources.")
+	flag.StringVar(&imagesBindAddr, "images-bind-addr", ":8084",
+		"The address the images endpoint binds to.")
+	flag.StringVar(&imagesPublishAddr, "images-publish-addr", "http://127.0.0.1:8084",
+		"The address clients would access the images endpoint from.")
+	flag.Parse()
+
+	ctrl.SetLogger(zap.New(zap.UseDevMode(devLogging)))
+
+	printVersion()
+
+	for _, env := range []string{"IRONIC_BASE_URL", "DEPLOY_ISO"} {
+		val := os.Getenv(env)
+		if val == "" {
+			setupLog.Info("Missing environment", "variable", env)
+			os.Exit(1)
+		}
+	}
+
+	_, err := url.Parse(imagesPublishAddr)
+	if err != nil {
+		setupLog.Error(err, "imagesPublishAddr is not parsable")
+		os.Exit(1)
+	}
+
+	imageServer := imagehandler.NewImageHandler(ctrl.Log.WithName("ImageHandler"), os.Getenv("DEPLOY_ISO"), imagesPublishAddr)
+	// why use a FileServer?
+	// 1. it streams files efficiently
+	// 2. if we cache these images, then that will be an easy change.
+	http.Handle("/", http.FileServer(imageServer.FileSystem()))
+	go func() {
+		log.Fatal(http.ListenAndServe(imagesBindAddr, nil))
+	}()
+
+	mgr, err := ctrl.NewManager(ctrl.GetConfigOrDie(), ctrl.Options{
+		Scheme:    scheme,
+		Port:      0, // Add flag with default of 9443 when adding webhooks
+		Namespace: watchNamespace,
```

Comment:
I'd suggest to add a cache selector for secrets here as we did in BMO to avoid memory issues when watching all the namespaces (https://github.com/metal3-io/baremetal-operator/blob/c0179b1507f975a59c5481faeee717d2876537f5/main.go#L135-L137)

Reply:
I was thinking that since it's just the network secret IINW, and since we have full control on its creation, we could directly create it with a specific label to allow defining a an ad-hoc selector here

Reply:
We only control its creation when it is done by the installer. If the user adds a Host later then this would be another manual step they'd have to do.

Reply:
Right, in any case if BMO is going to label it then there wouldn't be a big save.

Reply:
given that there is only one secret, i'll use ClientDisableCacheFor: []client.Object{&corev1.Secret{}},

Reply:
Sorry, missed this comment before posting a review. There's one Secret per resource, but there could be thousands of resources.

Reply:
More to the point, if we don't cache the Secrets then they don't get watched, so we don't get notified when they change.

Reply:
Agree, `ClientDisableCacheFor` was an option initially considered also for BMO but then discarded the due lack of notification